### PR TITLE
Use gid for static dynamic users

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/core/dynamic-user.c [hunk 1]
```diff
@@ -452,7 +457,7 @@ static int dynamic_user_realize(DynamicUser *d, char **suggested_dirs, uid_t *re
         if (r < 0)
                 return r;
 
-        r = dynamic_user_pop(d, &uid, &uid_lock_fd);
+        r = dynamic_user_pop(d, &num, &uid_lock_fd);
         if (r < 0) {
```
Comment (yuwata) [inline src/core/dynamic-user.c]: ~~The variable gid is not set when `dynamic_user_pop()` succeeds. So, needs to add~~
```
if (r >= 0)
        gid = num;
else {
```
I've proposed another simpler way.

File: src/core/dynamic-user.c [hunk 2]
```diff
@@ -435,12 +435,17 @@ static void unlockfp(int *fd_lock) {
...
 
         _cleanup_(unlockfp) int storage_socket0_lock = -1;
         _cleanup_close_ int uid_lock_fd = -1;
         _cleanup_close_ int etc_passwd_lock_fd = -1;
-        uid_t uid = UID_INVALID;
+        uid_t num = UID_INVALID; /* an uid if is_user, and a gid otherwise */
+        gid_t gid = GID_INVALID; /* a gid if is_user, ignored otherwise */
         int r;
 
         assert(d);
```
Comment (yuwata) [inline src/core/dynamic-user.c]:
```
assert(is_user == !!ret_uid);
assert(ret_gid);
```

File: src/core/dynamic-user.c [hunk 3]
```diff
@@ -530,22 +541,29 @@ static int dynamic_user_realize(DynamicUser *d, char **suggested_dirs, uid_t *re
...
          * dynamically right here, push that in along with the lock fd for it. */
-        r = dynamic_user_push(d, uid, uid_lock_fd);
+        r = dynamic_user_push(d, num, uid_lock_fd);
         if (r < 0)
                 return r;
 
-        *ret = uid;
+        if (is_user) {
+                *ret_uid = num;
+                *ret_gid = gid;
```
Comment (yuwata) [inline src/core/dynamic-user.c]: I think below is the simpler way.
`*ret_gid = gid == GID_INVALID ? num : gid;`

File: src/core/dynamic-user.c [hunk 4]
```diff
@@ -530,22 +541,29 @@ static int dynamic_user_realize(DynamicUser *d, char **suggested_dirs, uid_t *re
...
+        } else {
+                assert(ret_uid == NULL);
```
Comment (yuwata) [inline src/core/dynamic-user.c]: move to the beginning of this function.

File: src/core/dynamic-user.c [hunk 5]
```diff
@@ -530,22 +541,29 @@ static int dynamic_user_realize(DynamicUser *d, char **suggested_dirs, uid_t *re
                         /* Hmm, so as it appears there's now something stored in the storage socket. Throw away what we
                          * acquired, and use what's stored now. */
 
-                        unlink_uid_lock(uid_lock_fd, uid, d->name);
+                        unlink_uid_lock(uid_lock_fd, num, d->name);
                         safe_close(uid_lock_fd);
 
-                        uid = new_uid;
+                        num = gid = new_uid;
```
Comment (yuwata) [inline src/core/dynamic-user.c]: setting gid here is not necessary if you accept the change below.

File: src/core/dynamic-user.c [hunk 6]
```diff
@@ -502,26 +510,29 @@ static int dynamic_user_realize(DynamicUser *d, char **suggested_dirs, uid_t *re
...
+                if (num == UID_INVALID) {
                         /* No static UID assigned yet, excellent. Let's pick a new dynamic one, and lock it. */
 
-                        uid_lock_fd = pick_uid(suggested_dirs, d->name, &uid);
+                        uid_lock_fd = pick_uid(suggested_dirs, d->name, &num);
                         if (uid_lock_fd < 0)
                                 return uid_lock_fd;
+
+                        /* Use the same number for uid and gid (when is_user) */
+                        gid = num;
```
Comment (yuwata) [inline src/core/dynamic-user.c]: Also, setting gid here is not necessary if you accept the proposed change.

File: src/core/dynamic-user.c [hunk 7]
```diff
@@ -472,17 +477,20 @@ static int dynamic_user_realize(DynamicUser *d, char **suggested_dirs, uid_t *re
                         return etc_passwd_lock_fd;
 
                 /* First, let's parse this as numeric UID */
-                r = parse_uid(d->name, &uid);
-                if (r < 0) {
+                r = parse_uid(d->name, &num);
+                if (r >= 0)
+                        gid = num;
```
Comment (yuwata) [inline src/core/dynamic-user.c]: Also, setting gid here is not necessary if you accept the proposed change.

**Comment by keszybz** [comment]
Updated, with all the comments implemented.

**Comment by yuwata** [comment]
lgtm.

**Comment by yuwata** [comment]
The following CI failure in xenial-s390x seems not related.
```
E: Sub-process /usr/bin/dpkg returned an error code (1)
autopkgtest [14:32:05]: ERROR: erroneous package: Test dependencies are unsatisfiable. A common reason is that your testbed is out of date with respect to the archive, and you need to use a current testbed or run apt-get update or use -U.
blame: https://git.launchpad.net/~pitti/systemd/+git/debian
badpkg: Test dependencies are unsatisfiable. A common reason is that your testbed is out of date with respect to the archive, and you need to use a current testbed or run apt-get update or use -U.
```

**Comment by poettering** [comment]
lgtm

**Comment by poettering** [comment]
merging. CI looks good except of some packaging issues on s390x.

**Comment by martinpitt** [comment]
Nice, thanks @keszybz! Good to see CI being green again. The s390x failure is avahi-daemon sometimes failing in LXC, but that's indeed unrelated.

**Pull request merged**